fendermint/vm/interpreter/src/chain.rs

```diff
@@ -89,32 +97,72 @@ where
-    async fn process(&self, pool: Self::State, msgs: Vec<Self::Message>) -> anyhow::Result<bool> {
+    async fn process(&self, state: Self::State, msgs: Vec<Self::Message>) -> anyhow::Result<bool> {
         for msg in msgs {
-            if let ChainMessage::Ipc(IpcMessage::BottomUpExec(msg)) = msg {
-                let item = CheckpointPoolItem::BottomUp(msg);
-
-                // We can just look in memory because when we start the application, we should retrieve any
-                // pending checkpoints (relayed but not executed) from the ledger, so they should be there.
-                // We don't have to validate the checkpoint here, because
-                // 1) we validated it when it was relayed, and
-                // 2) if a validator proposes something invalid, we can make them pay during execution.
-                let is_resolved = atomically(|| match pool.get_status(&item)? {
-                    None => Ok(false),
-                    Some(status) => status.is_resolved(),
-                })
-                .await;
-
-                if !is_resolved {
-                    return Ok(false);
+            match msg {
+                ChainMessage::Ipc(IpcMessage::BottomUpExec(msg)) => {
+                    let item = CheckpointPoolItem::BottomUp(msg);
+
+                    // We can just look in memory because when we start the application, we should retrieve any
+                    // pending checkpoints (relayed but not executed) from the ledger, so they should be there.
+                    // We don't have to validate the checkpoint here, because
+                    // 1) we validated it when it was relayed, and
+                    // 2) if a validator proposes something invalid, we can make them pay during execution.
+                    let is_resolved = atomically(|| match state.0.get_status(&item)? {
+                        None => Ok(false),
+                        Some(status) => status.is_resolved(),
+                    })
+                    .await;
+
+                    if !is_resolved {
+                        return Ok(false);
+                    }
                 }
-            }
+                ChainMessage::Ipc(IpcMessage::TopDownProposal(ParentFinalityProposal {
+                    height,
+                    block_hash,
+                })) => {
+                    return if (atomically_or_err(|| {
+                        state.1.check_proposal(&IPCParentFinality {
+                            height: height as u64,
+                            // need clone as atomically_or_err seems not able to capture with move
+                            block_hash: block_hash.clone(),
+                        })
+                    })
+                    .await)
+                        .is_ok()
+                    {
+                        Ok(true)
+                    } else {
+                        Ok(false)
+                    };
```

Comment:
```suggestion
                    let prop = IPCParentFinality {
                            height: height as u64,
                            block_hash,
                    };
                    if atomically_or_err(|| state.1.check_proposal(&prop)).await.is_err() {
                      return Ok(false);
                    }
```
I think this is better because:
* It's more readable IMO. 
* You can create the prop once, because you only need a reference inside the tranaction, so you don't need to clone. (I think)
* You returned from the loop in both positive and negative outcomes. As you see from the bottom-up case, there can be other messages we need to check, so we can't return `true` just because the top-down is correct.